### PR TITLE
Convert arith.select to aievec.sel for different input data types.

### PR DESCRIPTION
**Pull request by linay-xsj**

### CODE REVIEW

**Review by david-vc — CHANGES_REQUESTED** [review]
HI @linay-xsj 
Can you please also add a test case with compare + select (like i16xi16_sel where the arith.cmpi has ult as predicate.) I think that it is important to show that we are generating correct C code also for this case.

**Comment by linay-xsj** [comment]
> HI @linay-xsj Can you please also add a test case with compare + select (like i16xi16_sel where the arith.cmpi has ult as predicate.) I think that it is important to show that we are generating correct C code also for this case.

Done.

File: lib/Targets/AIEVecToCpp/TranslateAIEVecToCpp.cpp [hunk 1]
```diff
@@ -1713,6 +1713,98 @@ static LogicalResult printOperation(CppEmitter &emitter,
...
+    os << "ge";
+  } else {
+    return failure();
+  }
+
+  os << "(";
+  VectorType vType = lhs.getType().cast<VectorType>();
+  Type eltType = vType.getElementType();
+
+  if (eltType.isa<IntegerType>() &&
```
Comment (david-vc) [inline lib/Targets/AIEVecToCpp/TranslateAIEVecToCpp.cpp:1756]: Slight variants of  piece of code 1758-1772 appears already multiple times in this file and so it should be refactored.

**Review by david-vc — COMMENTED** [review]
Thanks for adding the unsigned comparison + select test.
I have a minor request to refactor the code to build the name of a vector type given the element type and lane count.

**Review by david-vc — COMMENTED** [review]
Thanks for adding the unsigned comparison + select test.
I have a minor request to refactor the code to build the name of a vector type given the element type and lane count.

**Review by david-vc — APPROVED** [review]

File: lib/Targets/AIEVecToCpp/TranslateAIEVecToCpp.cpp [hunk 2]
```diff
@@ -1713,6 +1713,98 @@ static LogicalResult printOperation(CppEmitter &emitter,
...
+      (pred == "ult" || pred == "ule" || pred == "ugt" || pred == "uge")) {
+    unsigned lanes = getVectorLaneSize(vType);
+    unsigned width = getElementSizeInBits(vType);
+    os << "v" << std::to_string(lanes) << "uint" << std::to_string(width);
```
Comment (keryell) [inline lib/Targets/AIEVecToCpp/TranslateAIEVecToCpp.cpp:1760]: It sounds surprising that the streams do not accept directly integers.

File: include/aie/Dialect/AIEVec/IR/AIEVecOps.td [hunk 3]
```diff
@@ -696,4 +696,20 @@ def AIEVec_CmpOp:
...
+def AIEVec_SelOp:
+  AIEVec_Op<"sel", [
+    Pure,
+    AllTypesMatch<["lhs", "rhs", "result"]>
+  ]>,
+  Arguments<(ins AnyVector:$lhs, AnyVector:$rhs, AnyUnsignedInteger:$sel)>,
+  Results<(outs AnyVector:$result)> {
+  let summary = "AIE vector lane wise selection";
+  let description = [{
+    Xilinx-specific intrinsic that performs lane wise selection between two input vectors, if a bit of sel is zero, the lane of vector lhs is selected, else the lane of vector rhs is selected.
```
Comment (keryell) [inline include/aie/Dialect/AIEVec/IR/AIEVecOps.td:709]: Replacing Xilinx by AMD in new code?

File: test/unit_tests/aievec_tests/bf16xbf16_sel/defines.h [hunk 4]
```diff
@@ -0,0 +1,4 @@
+#pragma once
+#define IN0_SIZE 1024
+#define IN1_SIZE 1024
+#define OUT0_SIZE 1024
```
Comment (keryell) [inline test/unit_tests/aievec_tests/bf16xbf16_sel/defines.h:4]: Any reason for not using `constexpr` variables?

File: test/unit_tests/aievec_tests/bf16xbf16_sel/testbench.cc [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+alignas(32) bfloat16 g_out0[OUT0_SIZE];
+alignas(32) bfloat16 g_out0Ref[OUT0_SIZE];
+
+int main(int argc, char *argv[]) {
+  std::string dataDir(TO_STR(DATA_DIR));
+  srand(10);
+  std::generate(g_in0, g_in0 + IN0_SIZE,
+                [&]() mutable { return random_bfloat16(-10, 10, 2); });
+  std::generate(g_in1, g_in1 + IN1_SIZE,
+                [&]() mutable { return random_bfloat16(-10, 10, 2); });
```
Comment (keryell) [inline test/unit_tests/aievec_tests/bf16xbf16_sel/testbench.cc:22]: Unclear to me why this is `mutable`.

File: test/unit_tests/aievec_tests/floatxfloat_sel/defines.h [hunk 6]
```diff
@@ -0,0 +1,4 @@
+#pragma once
+#define IN0_SIZE 1024
```
Comment (keryell) [inline test/unit_tests/aievec_tests/floatxfloat_sel/defines.h:2]: https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#es31-dont-use-macros-for-constants-or-functions

**Pull request merged**